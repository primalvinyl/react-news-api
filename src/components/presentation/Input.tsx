import React from 'react';
import styles from './Input.module.scss';

const Input = React.forwardRef<HTMLInputElement, InputType>(
    (props, ref) => {
    
    const {
        id,
        value,
        className,
        name,
        label,
        placeholder,
        type,
        required,
        disabled,
        errors, 
        touched,
        handleChange,
        handleBlur
    } = props;

    return (
        <div className={styles.root}>
            {(label) && (<label htmlFor={id}>
                {label}
            </label>)}
            <div className={styles.formControlWrapper}>
                <input
                    type={type}
                    className={className}
                    id={id}
                    name={name}
                    value={value}
                    ref={ref}
                    placeholder={placeholder}
                    required={required}
                    disabled={disabled}
                    onChange={handleChange}
                    onBlur={handleBlur}
                />
                {(errors && touched) && (errors[id] && touched[id]) && (
                    <div className={styles.formFieldError}>
                        {errors[id]}
                    </div>
                )}
            </div>
        </div>
    );
});

type InputType = {
    readonly id: string;
    readonly value: string;
    readonly className?: string;
    readonly name?: string;
    readonly label?: string;
    readonly placeholder?: string;
    readonly type?: string;
    readonly required?: boolean;
    readonly disabled?: boolean;
    readonly errors?: { [key: string]: string };
    readonly touched?: { [key: string]: string };
    readonly handleChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
    readonly handleBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
};

Input.defaultProps = {
    className: undefined,
    name: undefined,
    label: undefined,
    placeholder: undefined,
    type: 'text',
    required: false,
    disabled: false,
    errors: {},
    touched: {},
    handleChange: (): void => {},
    handleBlur: (): void => {}
}

export default Input;
