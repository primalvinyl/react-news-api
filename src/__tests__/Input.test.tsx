import * as React from 'react';
import { shallow } from 'enzyme';
import Renderer from 'react-test-renderer';
import Input from '../components/presentation/Input';

describe('Input', () => {
    it('renders', () => {
        const myWrapper = shallow(
            <Input
                id="test"
                value="test" />
        );
        expect(myWrapper.exists());
    });
    it('matches snapshot', () => {
        const component = Renderer.create(
            <Input
                id="test"
                value="test" />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});