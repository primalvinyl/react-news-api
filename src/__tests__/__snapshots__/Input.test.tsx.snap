// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input matches snapshot 1`] = `
<div
  className="root"
>
  <div
    className="formControlWrapper"
  >
    <input
      disabled={false}
      id="test"
      onBlur={[Function]}
      onChange={[Function]}
      required={false}
      type="text"
      value="test"
    />
  </div>
</div>
`;
